File: exo4.c
#include <stdio.h>
#include <stdlib.h>

/// \param n le nombre de boites
/// \param valeur la valeur du marché du jour
/// \param boites le contenu de chaque boite
int le_juste_etal(int n, int valeur, int* boites) {

	int* cache = malloc(n*sizeof(int));
	int count = 0;

	for(int i = 0; i < n; i++) 
		for(int j = 0; j <= i; j++)
			if((cache[j]+=boites[i]) % valeur == 0)count++;

	free(cache);
	return count % 1000000007;
}

int main() {
    int n;
    scanf("%d", &n);
    int valeur;
    scanf("%d", &valeur);
    int* boites = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; ++i)
        scanf("%d", &boites[i]);
    
    int e = le_juste_etal(n, valeur, boites);
	printf("%i \n", e);
    return 0;
}

File: exo4.ml
(**
   @param n le nombre de boites
   @param valeur la valeur du marché du jour
   @param boites le contenu de chaque boite
*)
let rec calc v boites =
  (** TODO Afficher, sur une ligne, un unique entier : le nombre de séries
  justes modulo $1,000,000,007$.  *)

        match boites with 
                | [] -> ()
                | h::q -> begin
                        calc v q;
                        calc v q
                end
;;

let () =
  let n = read_int () in
  let boites = read_line () |> fun x -> if x = "" then [] else String.split_on_char ' ' x |> List.rev_map int_of_string |> List.rev in
  let r = etal n valeur boites in
        Printf.printf "%i \n" r;;
